streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredWindowStore.java

```diff
@@ -292,13 +408,46 @@ public V fetch(final K key,
             time);
     }
 
+    private long getActualWindowStartTime(final long timeFrom) {
+        return Math.max(timeFrom, ((PersistentWindowStore<Bytes, byte[]>) wrapped()).getObservedStreamTime() - retentionPeriod + 1);
+    }
+
+    private KeyValueIterator<Windowed<K>, V> filterExpiredRecords(final boolean forward) {
+        final KeyValueIterator<Windowed<Bytes>, byte[]> allWindowedKeyValueIterator = forward ? wrapped().all() : wrapped().backwardAll();
+
+        final long observedStreamTime = ((PersistentWindowStore<Bytes, byte[]>) wrapped()).getObservedStreamTime();
+        if (!allWindowedKeyValueIterator.hasNext() || observedStreamTime == ConsumerRecord.NO_TIMESTAMP)
+            return new MeteredWindowedKeyValueIterator<>(allWindowedKeyValueIterator, fetchSensor, streamsMetrics, serdes, time);
+
+        final long windowStartBoundary = observedStreamTime - retentionPeriod + 1;
+        final List<KeyValue<Windowed<Bytes>, byte[]>> windowedKeyValuesInBoundary = new ArrayList<>();
+
+        while (allWindowedKeyValueIterator.hasNext()) {
+            final KeyValue<Windowed<Bytes>, byte[]> next = allWindowedKeyValueIterator.next();
+            if (next.key.window().endTime().isBefore(Instant.ofEpochMilli(windowStartBoundary))) {
+                continue;
+            }
+            windowedKeyValuesInBoundary.add(next);
+        }
+        return new MeteredWindowedKeyValueIterator<>(new WindowedKeyValueIterator(windowedKeyValuesInBoundary.iterator()), fetchSensor, streamsMetrics, serdes, time);
+    }
```

Comment:
@mjsax i have a question here... In the jira ticket, you. have mentioned that the best place for adding this filtering is in the MeteredStore as that implicitly adds the logic even for custom state stores. While for the most part, this kind of filtering has worked fine(fetching relevant records and then filtering in MeteredStore) but there's a case where it's failing. It's for test. cases like `shouldNotThrowConcurrentModificationException` . This seems to be because the put() call while iterating is appending to the wrapped instance of iterator and hence it's not visible.

Looking at this, do you think it would be a good idea to move this logic in the actual RocksDB implementations? Or do you think there's a better way to do it here in MeteredStore class itself?

Reply:
> @mjsax i have a question here... In the jira ticket, you. have mentioned that the best place for adding this filtering is in the MeteredStore as that implicitly adds the logic even for custom state stores. While for the most part, this kind of filtering has worked fine(fetching relevant records and then filtering in MeteredStore) but there's a case where it's failing. It's for test. cases like `shouldNotThrowConcurrentModificationException` . This seems to be because the put() call while iterating is appending to the wrapped instance of iterator and hence it's not visible.
> 
> Looking at this, do you think it would be a good idea to move this logic in the actual RocksDB implementations? Or do you think there's a better way to do it here in MeteredStore class itself?

@showuon / @ableegoldman would like to know your thoughts as well on this..

Reply:
hey.. let me know what you guys think about the above comment of mine. Based upon that, I will proceed with fixing the issue with the test case(which doesn't seem possible with the way I have implemented).

Reply:
>  it's failing for test cases like shouldNotThrowConcurrentModificationException . This seems to be because the put() call while iterating....

You don't mean that the test itself is failing, just that the filter isn't being applied to the `put` records, right?
> ...the put() call while iterating is appending to the wrapped instance of iterator and hence it's not visible

Can you also expand a bit on what you mean by this? Not sure I have the whole picture here since I haven't been following this too closely, thanks in advance for catching me up 🙂  

> Looking at this, do you think it would be a good idea to move this logic in the actual RocksDB implementations? Or do you think there's a better way to do it here in MeteredStore class itself?

Personally I think the benefits of keeping this in the metering layer outweigh any downsides, especially if it comes down to just some weird edge case(s) that don't play nicely with the filtering. But I'm also interested in hearing what @showuon makes of this

Reply:
@ableegoldman I agree that having the filtering logic in the Metering layer is the most sensible place to keep it as that would provide the flexibility to add this filtering logic for any store. 
Let me explain the scenario with `shouldNotThrowConcurrentModificationException`:

If you see the test case in AbstractWindowBytesStoreTest:

```
try (final KeyValueIterator<Windowed<Integer>, String> iterator = windowStore.all()) {

            currentTime += WINDOW_SIZE * 10;
            windowStore.put(1, "three", currentTime);

            currentTime += WINDOW_SIZE * 10;
            windowStore.put(2, "four", currentTime);

            // Iterator should return all records in store and not throw exception b/c some were added after fetch
            assertEquals(windowedPair(1, "one", 0), iterator.next());
            assertEquals(windowedPair(1, "two", WINDOW_SIZE * 10), iterator.next());
            assertEquals(windowedPair(1, "three", WINDOW_SIZE * 20), iterator.next());
            assertEquals(windowedPair(2, "four", WINDOW_SIZE * 30), iterator.next());
            assertFalse(iterator.hasNext());
        }
```

after fetching all the keys present in the store using `all()`, during iteration, 2 new keys are being `put`. With the current implementation in trunk, it works fine but with the filtering logic that I have added for all() in `filterExpiredRecords` : 

```
private KeyValueIterator<Windowed<K>, V> filterExpiredRecords(final boolean forward) {
        final KeyValueIterator<Windowed<Bytes>, byte[]> allWindowedKeyValueIterator = forward ? wrapped().all() : wrapped().backwardAll();
        final long observedStreamTime = getObservedStreamTime(wrapped());
        if (!allWindowedKeyValueIterator.hasNext() || observedStreamTime == ConsumerRecord.NO_TIMESTAMP)
            return new MeteredWindowedKeyValueIterator<>(allWindowedKeyValueIterator, fetchSensor, streamsMetrics, serdes, time);

        final long windowStartBoundary = observedStreamTime - retentionPeriod + 1;
        final List<KeyValue<Windowed<Bytes>, byte[]>> expiredRecords = new ArrayList<>();

        while (allWindowedKeyValueIterator.hasNext()) {
            final KeyValue<Windowed<Bytes>, byte[]> next = allWindowedKeyValueIterator.next();
            if (next.key.window().endTime().isBefore(Instant.ofEpochMilli(windowStartBoundary))) {
                expiredRecords.add(next);
            }
        }
        for (KeyValue<Windowed<Bytes>, byte[]> record: expiredRecords) {
            wrapped().put(record.key.key(), null, record.key.window().start());
        }
        return new MeteredWindowedKeyValueIterator<>(wrapped().all(), fetchSensor, streamsMetrics, serdes, time);
//        return new MeteredWindowedKeyValueIterator<>(new WindowedKeyValueIterator(expiredRecords.iterator()), fetchSensor, streamsMetrics, serdes, time);
    }
```

first `all` the keys are fetched and then the filtering happens afterwards. At this point, if you notice, I am essentially. creating a new iterator which is what being returned. Now, in the failing test case, when `put` inserts 2 more keys during iteration, that happens to the original iterator in `wrapped()` and not to the iterator being returned. And that's the reason, the test case fails on

`assertEquals(windowedPair(1, "three", WINDOW_SIZE * 20), iterator.next());` as there are no more records. Just for testing, if I uncomment the last return and run the test. case again, it works fine.

So, either there's something wrong with the way I have implemented, where any suggestions would be welcome or due to the edge case, it might be needed to push the filtering down further to the underlying store. Let me know if it made sense.

Reply:
hey @showuon , did you get a chance to look at my comments above?

Reply:
I'll take a look next week. Thanks.

Reply:
@vamossagar12 , sorry for late reply. I read through your changes again, and I'm thinking that if we could fetch a specific time range via `fetch` API, instead of `fetchAll`, then doing manually filtering?

For example:
```java
private KeyValueIterator<Windowed<K>, V> filterExpiredRecords(final boolean forward) {
        // this is what you've done, to fetch all, then filter later.
        // final KeyValueIterator<Windowed<Bytes>, byte[]> allWindowedKeyValueIterator = forward ? wrapped().all() : wrapped().backwardAll();
       
       final long observedStreamTime = getObservedStreamTime(wrapped());
       if (!allWindowedKeyValueIterator.hasNext() || observedStreamTime == ConsumerRecord.NO_TIMESTAMP)
            return new MeteredWindowedKeyValueIterator<>(allWindowedKeyValueIterator, fetchSensor, streamsMetrics, serdes, time);


        
        
     

       final long windowStartBoundary = observedStreamTime - retentionPeriod + 1;

       // what about this? (use fetchAll(Instant timeFrom, Instant timeTo) to fetch the records in time range, by setting timeTo to the boundary we computed)
       final KeyValueIterator<Windowed<Bytes>, byte[]> allWindowedKeyValueIterator = forward ? wrapped().fetchAll(0, windowStartBoundary) : wrapped().backwardFetchAll(0, windowStartBoundary);

       return new MeteredWindowedKeyValueIterator<>(allWindowedKeyValueIterator, fetchSensor, streamsMetrics, serdes, time);

        


        /*
        final long windowStartBoundary = observedStreamTime - retentionPeriod + 1;
        final List<KeyValue<Windowed<Bytes>, byte[]>> expiredRecords = new ArrayList<>();

        while (allWindowedKeyValueIterator.hasNext()) {
            final KeyValue<Windowed<Bytes>, byte[]> next = allWindowedKeyValueIterator.next();
            if (next.key.window().endTime().isBefore(Instant.ofEpochMilli(windowStartBoundary))) {
                expiredRecords.add(next);
            }
        }
        for (KeyValue<Windowed<Bytes>, byte[]> record: expiredRecords) {
            wrapped().put(record.key.key(), null, record.key.window().start());
        }
       */

        //return new MeteredWindowedKeyValueIterator<>(wrapped().all(), fetchSensor, streamsMetrics, serdes, time);
//        return new MeteredWindowedKeyValueIterator<>(new WindowedKeyValueIterator(expiredRecords.iterator()), fetchSensor, streamsMetrics, serdes, time);
    }
```


Does that make sense?

Reply:
@showuon that's an interesting idea. But, I think it needs to be in the opposite order ie, we need to fetch from windowStartBoundary -> end. We don't know the end timestamp in this case. (The above code is wrong as I checked in during some testing).

So, what I mean to say is, 

this:

`final KeyValueIterator<Windowed<Bytes>, byte[]> allWindowedKeyValueIterator = forward ?
 wrapped().fetchAll(0, windowStartBoundary) : wrapped().backwardFetchAll(0, windowStartBoundary);`

should be 

`final KeyValueIterator<Windowed<Bytes>, byte[]> allWindowedKeyValueIterator = forward ?
 wrapped().fetchAll(windowStartBoundary, end) : wrapped().backwardFetchAll(windowStartBoundary, end);`


But, we don't know the end. I ran it with Long.MAX_VALUE and the test case in question seems to have passed. Thanks for that!

I will clean up this code again and send for review again. Thanks again!

Reply:
Awesome!